Configure mesh coordinate system

orsonbraines opened the pull request:
## Description
- Implements #2322 
- Add a new combo box and button to camera settings for mesh viewer
- Add a new dialog window for custom axis mappings
- Implement the axis mappings in gl driver (other drivers to come)

orsonbraines committed:
Issue 2322: Configure mesh coordinate system

- Add a new combo box and button to camera settings for mesh viewer
- Add a new dialog window for custom axis mappings
- Implement the axis mappings in gl driver (other drivers to come)

orsonbraines commented on renderdoc/driver/gl/gl_debug.cpp:
```diff
@@ -2238,6 +2238,21 @@ bool GLReplay::GetHistogram(ResourceId texid, const Subresource &sub, CompType t
   return true;
 }
 
+static Matrix4f getAxisMapMat(const MeshDisplay &cfg)
```
I wasn't sure of a good place to put this logic so we could use it across all drivers. Any suggestions?

baldurk replied:
If you made a sub-struct to contain `xAxisMapping`, `yAxisMapping`, and `zAxisMapping` you could make a new `Matrix4f` constructor that took that struct as a parameter. If you look at how `Vec4f` has a `FloatVector` constructor, it forward declares that struct in the header then includes `api/replay/data_types.h` in the source - you could do a similar thing.

baldurk commented on qrenderdoc/Windows/BufferViewer.cpp:
```diff
@@ -2945,7 +2950,18 @@ void BufferViewer::populateBBox(PopulateBufferData *bufdata)
 {
-  QVariantMap state = ui->dockarea->saveState();
+  QVariantMap state;
+  state[lit("dockarea")] = ui->dockarea->saveState();
```
Changing where the dock area state is saved like this will mean that users will lose their custom layout after this change is merged.

You should leave the `QVariantMap` as-is when returned from the dockarea like this, and just add your custom entries to the root of it. Any extra data will be ignored by `dockarea->restoreState` as it only looks for the keys it cares about.

baldurk commented on qrenderdoc/Windows/BufferViewer.cpp:
```diff
@@ -4030,6 +4059,57 @@ void BufferViewer::camGuess_changed(double value)
+  else
+  {
+    ui->axisMappingButton->setEnabled(true);
+  }
```
I'm thinking that we should have it so that newly selecting the custom index immediately pops up the axis mapping dialog, and if that dialog is cancelled it should revert back to the previously selected item. That prevents the situation where a user selects the `Custom...` entry but then doesn't customise the axis mapping, so it will invisibly use whatever was previously there and the UI is a bit inconsistent with what's actually happening.

Maybe this is annoying in practice but I feel like 95% of the time when you select custom you're going to want to then customise it, so this also saves the user a click?

baldurk commented on qrenderdoc/Windows/Dialogs/AxisMappingDialog.cpp:
```diff
@@ -0,0 +1,156 @@
+  ui->setupUi(this);
+  setWindowFlags(windowFlags() & ~Qt::WindowContextHelpButtonHint);
+
+  const QStringList items{tr("Right"), tr("Left"),        tr("Up"),
```
Please use `items = { ... }` for initialisation.

baldurk commented on qrenderdoc/Windows/Dialogs/AxisMappingDialog.cpp:
```diff
@@ -0,0 +1,156 @@
+  {
+    QMessageBox messageBox;
+    messageBox.critical(0, tr("Error"), tr("Your axis mappings are not compatible."));
+    messageBox.setFixedSize(700, 150);
```
Please use `RDDialog::critical` for this just so that any project-wide customisations are applied. E..g. there are some issues with using Qt's event looks from the base code on windows which requires the use of `RDDialog::show`.

You should also expand on the error message - explain to the user that the axis mapping is degenerate and doesn't cover all three directions. Ideally you could point out exactly which two are the problem, e.g. "X and Y are both mapped to the left-right axis" or something like that.

baldurk reviewed (changes requested):
Overall this looks good I think. I've left some notes on some bits of the code about changes to make.

One thing I noticed from testing this myself is that the default camera placement (at -Z from the AABB centre looking forward) breaks when custom axis mapping is used. I think the matrix multiplication order is fine as the camera controls are correct after that but I think you'll have to account for the axis mapping when placing/resetting the camera.

jskattuk committed:
address PR feedback

- create AxisMapping struct and fix camera location issue
- add more detail to axis mapping error message
- fix other minor issues

orsonbraines commented:
@baldurk  We've now addressed all of the feedback. Ready for rereview. Also, do you want us to work on applying the custom mappings in the other drivers now and add them to this PR or tackle those in a follow up PR?

baldurk commented:
This looks really good now! Having tried it, I think the behaviour of immediately popping up the axis dialog on selecting 'Custom' feels right, would you agree?

The only thing is that since you've removed the explicit button to edit the mapping, there's no way for the user to go back into fix a mistake if they selected the wrong thing without selecting a different axis mapping and then back to custom. That's also the only way currently to view what the custom mapping is - which I think is fine, but it needs a way to go back to that dialog. Even if we allowed them to "re-select" custom to pop up the dialog, that's not very intuitive. I think it would be good to keep the button you had previously, but still preserve this behaviour where it automatically pops up when selecting custom.

One other minor note - similar to how the matrix parameter inputs are disabled when looking at the VS input (because they're not used), we should probably disable the axis mapping drop-down/button when looking at the VS output. It doesn't do much harm but it could confuse people if they expect it to work and don't understand why it has no effect.

The CI failure is a check to make sure all python structs/functions are listed in the documentation - not very critical for structs like this but it's better to have it listed. If you look in `docs/python_api/renderdoc/outputs.rst` I think the `AxisMapping` struct would fit best next to the items at the bottom under the `Mesh View` heading.

baldurk commented:
> Also, do you want us to work on applying the custom mappings in the other drivers now and add them to this PR or tackle those in a follow up PR?

Sorry I meant to respond to this too - that's up to you really, I don't require strict feature parity in PRs like this where it's not a significant burden to handle it myself, so after the above issues are resolved I'm happy to merge this as-is. If this were a large scale feature that I would be unlikely to implement myself on other APIs, I wouldn't expect support for all APIs in a single PR as it's fine to add things piece-wise but I would expect follow up PRs for the other APIs at some stage.

I would certainly very much appreciate you adding support in the other drivers, as I'd want that before I close the linked issue, but whether you want to keep working on that in this PR or do a follow-up, or even do neither, any of those options are fine. Just let me know what you plan on doing so I can extend the support myself if needed :smile:.

jskattuk committed:
readded custom axis mapping button and updated documentation

jskattuk commented:
@baldurk We've made the changes to address your latest review comments. We've also decided to handle custom mappings for the other drivers in a follow up PR.

baldurk commented:
Thanks, I've merged now. Just a heads up - I'd appreciate a PR for the other APIs within the next 2 weeks or so, otherwise I'll likely look at implementing it myself. The next stable version will be coming out at the end of this month and I think it makes more sense to ensure the other API implementations are complete - rather than either having confusingly non-functional UI or setting up something to make it conditionally disabled/enabled depending on the underlying API.
